File: lib/core/config/tenant.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// The church the app is currently acting as.
///
/// This is the single value that turns one binary into any church's app.
/// Every Firestore repository is built from it (see `lib/app/backend.dart`),
/// so changing it rebuilds the whole data layer and the app re-themes and
/// re-reads without a restart.
///
/// Overridden at startup with whatever was last chosen, and written by the
/// church picker. Null means nobody has chosen yet, which the router turns
/// into the picker.
final selectedChurchIdProvider = StateProvider<String?>((ref) => null);

/// The church id repositories should use right now.
///
/// Falls back to [demoChurchId] rather than throwing: the zero-backend
/// build has no picker to go through, and a null here would otherwise
/// mean every repository had to handle "no church yet" separately.
final currentChurchIdProvider = Provider<String>((ref) {
  return ref.watch(selectedChurchIdProvider) ?? demoChurchId;
});

/// The church the bundled sample content describes.
///
/// Also the id a single-church deployment can use if it never wants a
/// picker at all - point every member at this one and the app behaves
/// exactly as it did before it learned about tenancy.
const String demoChurchId = 'yoked-demo';

/// The prefix every church-scoped route carries.
///
/// Short on purpose: it is in front of every URL a church will ever
/// print on a card or read out from the front.
const String churchPathPrefix = '/c';

/// The address of a page within a church.
///
/// `churchPath('riverside', '/sermons')` is `/c/riverside/sermons`. The
/// one place that shape is written down, so changing the prefix is a
/// one-line change rather than a search across the app.
String churchPath(String churchId, [String subPath = '/']) {
  final rest = subPath == '/' ? '' : subPath;
  return '$churchPathPrefix/$churchId$rest';
}

/// A church's whole address, as somebody would write it on a card.
///
/// [churchPath] is the part the router reasons about; this is the part
/// you can hand out. Built from [Uri.base], which on the web is the page
/// URL, so it picks up the real origin and the deploy's base path -
/// `/Yoked-Church-Web-App/` here, `/` for a fork on its own domain -
/// rather than being hardcoded to one deployment.
///
/// The `#` is included because that is the address the router answers
/// to. `web/404.html` makes it optional for anyone typing it in, but the
/// version offered for copying should be the one that works everywhere,
/// including when it is pasted somewhere that does not follow redirects.
///
/// Returns empty where there is no web address to give: a desktop or
/// mobile build's [Uri.base] is a `file:` directory with no origin, and
/// inventing one would put a URL that goes nowhere on a copy button.
String churchUrl(String churchId, {Uri? from}) {
  final base = from ?? Uri.base;
  if (base.scheme != 'http' && base.scheme != 'https') return '';

  // The directory the app is served from. A page served as
  // `.../index.html` has that on the end, and it is not part of the
  // address anyone should be given.
  var dir = base.path;
  if (!dir.endsWith('/')) dir = dir.substring(0, dir.lastIndexOf('/') + 1);

  return '${base.origin}$dir#${churchPath(churchId)}';
}

/// The church a location names, or null if it names none.
///
/// Pure, and used from two places that cannot share anything else:
/// `main()` reads it from the browser URL *before the first frame* so a
/// deep-linked visitor never sees another church's content flash past,
/// and the router reads it from every navigation afterwards.
String? churchIdFromLocation(String location) {
  final segments = Uri.parse(location).pathSegments;
  if (segments.length < 2 || '/${segments.first}' != churchPathPrefix) return null;
  final id = segments[1].trim();
  return id.isEmpty ? null : id;
}

/// Strips the church prefix, leaving the path the app reasons about.
///
/// Every route guard, feature flag and role check was written against
/// bare paths like `/admin/settings` and stays that way; only this
/// function knows the difference.
String subPathOf(String location) {
  final id = churchIdFromLocation(location);
  if (id == null) return location;
  final rest = location.substring('$churchPathPrefix/$id'.length);
  return rest.isEmpty ? '/' : rest;
}

File: src/core/tenant.ts
/**
 * Which church the app is acting as.
 *
 * One value turns one bundle into any church's app: every repository is
 * built from it, so changing it re-reads the whole data layer and
 * re-themes the storefront without a reload.
 *
 * Ported from `lib/core/config/tenant.dart` with one deliberate change,
 * noted on `churchUrl` below.
 */

/** The church the bundled sample content describes. */
export const DEMO_CHURCH_ID = 'yoked-demo';

/**
 * The prefix every church-scoped route carries.
 *
 * Short on purpose: it sits in front of every URL a church will ever
 * print on a card or read out from the front.
 */
export const CHURCH_PATH_PREFIX = '/c';

/**
 * The address of a page within a church.
 *
 * `churchPath('riverside', '/giving')` is `/c/riverside/giving`. The one
 * place that shape is written down, so moving the prefix stays a
 * one-line change rather than a search across the app.
 */
export function churchPath(churchId: string, subPath = '/'): string {
  const rest = subPath === '/' ? '' : subPath;
  return `${CHURCH_PATH_PREFIX}/${churchId}${rest}`;
}

/**
 * A church's whole address, as somebody would write it on a card.
 *
 * **No `#`.** The Flutter build had to include one — it routed on the
 * fragment, so `example.com/c/riverside` reached the server as
 * `example.com` and the church was lost. React Router uses real paths,
 * and `public/404.html` bounces a deep link back through the SPA with
 * the path intact, so the address a church hands out is now the address
 * a browser shows.
 *
 * The deploy's base path comes from `import.meta.env.BASE_URL`, which is
 * the `base` Vite was built with — `/Yoked-Church-Web-App/` here, `/`
 * for a fork on its own domain.
 *
 * Deriving it from the current pathname instead, which is what the Dart
 * version did, is wrong now and was only ever right by accident. Under
 * hash routing the pathname was always the base; under real paths it is
 * the page you are standing on, so an admin who opened this from
 * `/c/riverside/settings` would have been handed
 * `example.com/c/riverside/c/riverside` to print on a card.
 *
 * Returns empty where there is no web address to give, rather than
 * inventing one and putting a dead URL behind a copy button.
 */
export function churchUrl(
  churchId: string,
  from: { origin: string; protocol: string } = window.location,
  basePath: string = import.meta.env.BASE_URL,
): string {
  if (from.protocol !== 'http:' && from.protocol !== 'https:') return '';

  // `BASE_URL` always carries a trailing slash; `churchPath` always
  // leads with one, and two would give `//c/riverside`.
  const dir = basePath.endsWith('/') ? basePath.slice(0, -1) : basePath;

  return `${from.origin}${dir}${churchPath(churchId)}`;
}

/** The church a location names, or null if it names none. */
export function churchIdFromLocation(location: string): string | null {
  const segments = location.split('?')[0]!.split('#')[0]!.split('/').filter(Boolean);
  if (segments.length < 2 || `/${segments[0]}` !== CHURCH_PATH_PREFIX) return null;
  const id = segments[1]!.trim();
  return id === '' ? null : id;
}

/**
 * Strips the church prefix, leaving the path the app reasons about.
 *
 * Every guard, feature flag and role check is written against bare paths
 * like `/admin/settings` and stays that way; only this function knows
 * the difference.
 */
export function subPathOf(location: string): string {
  const id = churchIdFromLocation(location);
  if (id === null) return location;
  const rest = location.slice(`${CHURCH_PATH_PREFIX}/${id}`.length);
  return rest === '' ? '/' : rest;
}
